File: Grid/Spiral Matrix/54. Spiral Matrix.py
#problem link: https://leetcode.com/problems/spiral-matrix/

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m, n = len(matrix), len(matrix[0])
        
        top, bottom = 0, m - 1
        left, right = 0, n - 1
        
        res = []
        
        while top <= bottom and left <= right:
            # left -> right
            for i in range(left, right + 1):
                res.append(matrix[top][i])
            top += 1
                
            # top -> bottom
            for i in range(top, bottom + 1):
                res.append(matrix[i][right])
            right -= 1
                
            # right -> left
            if top <= bottom:
                for i in range(right, left - 1, -1):
                    res.append(matrix[bottom][i])
                bottom -= 1
                
            # bottom -> top
            if left <= right:
                for i in range(bottom, top -1, -1):
                    res.append(matrix[i][left])
                left += 1
        return res

File: Grid/Spiral Matrix/54. Spiral Matrix.cpp
//problem link: https://leetcode.com/problems/spiral-matrix/

#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int n = matrix.size(), m = matrix[0].size();
        int l = 0, top = 0;
        int bottom = n - 1, r = m - 1;
        vector<int> res;
        while (l <= r && top <= bottom) {

            // left to right
            for (int i = l; i <= r; i++) {
                res.push_back(matrix[top][i]);
            }
            top++;

            // top to bottom
            for (int i = top; i <= bottom; i++) {
                res.push_back(matrix[i][r]);
            }
            r--;

            // right to left 
            if (top <= bottom){
                for (int i = r; i >= l; i--) {
                    res.push_back(matrix[bottom][i]);
                }
                bottom--;
            }

            // bottom to top
            if (l <= r) {
                for (int i = bottom; i >= top; i--) {
                    res.push_back(matrix[i][l]);
                }
                l++;
            }
        }
        return res;
    }
};